### DISCUSSION TITLE
queryFeatures deprecation

### DISCUSSION THREAD

**Discussion by eliboni** [protomaps/protomaps.js#143] [General]
I've been surprised to discover `queryFeatures`  has been deprecated. I've released an updated version of my "hobby" project  (mappite.org)  that relies on that feature to do something exceptional that was unthinkable for a small side project to achieve before "the single most exciting development in digital mapping in the past 10 years" (which I really think protomaps is!): I could build pmtiles  with specific road features (in my case, unpaved roads) and add a layer to let users to know road details (via queryFeature) and link to openstreetmap for corrections. Giving visibility to road details is the basis to improve the map. For my purpose, it worked like a charm.

Although I undertstand queryFeatures may have some limitations, I wander if it can be considered to pull it back, perhaps with a clear disclaimer of how it can be used, since it really offered a plus over standard leaflet layers or raster maps.

Moving to maplibre is a long path for apps built on leaflet, and perhaps not a real advantage for the purpose. I can stick with v1.24 but  it is clearly a risk going forward.

Pls. let me know your thoughts! 

... and in anycase thanks for the great work and the innovation protomaps brings to all of us.

**Comment by bdon** [comment]
Where on mappite.org is this used? There are major downsides to how `queryFeatures` works:

1.  There is no way to change the mouse cursor when hovering over an interactable feature
2. It is complicated and slow to change the appearance of a selected/hovered feature
3. The return of `queryFeatures` is the data in the tiles as a whole, not limited to what is rendered

The above are mostly constraints of how HTML Canvas rendering works. if you use plain `L.geoJSON` or a SVG/DOM based rendering method the above becomes easier.

While it is technically possible to implement everything, it is redundant work with MapLibre and not a good use of limited time.

**Comment by eliboni** [comment]
> Where on mappite.org is this used? ...

It is explained here: https://www.advrider.com/f/threads/mappite-org-create-routes-easily-feedback-wanted.1055040/page-12#post-49712507 
Zoom in while in USA or over Alps, a toast will inform the area has the "beta" Unpaved Roads highlight feature, on right menu the unpaved road icon will appear. The result is an "augmented" view on top of OSM tiles (brown lines with red/green border depending on vehicle access), right click on an unpaved road to get road details and link to OSM object:

![5-Qwj8CLo](https://github.com/protomaps/protomaps-leaflet/assets/17626755/3b222518-8288-4e2a-a412-29ada884bab0)

![2-kEL7MVF](https://github.com/protomaps/protomaps-leaflet/assets/17626755/9c80f66b-68b4-4db6-a1a4-7e9b7eb83a32)

in my use case, issue 1&2 are not relevant since I don't need to highlight a specific feature and since the app is mainly used from mobile screens the mouse cursor change is not required. Point 3 is not relevant as well since  alla features appears from a certain zoom (just the color/size differs). I understand these can be limitations for other projects.

**Reply by bdon** [reply]
What alternate approaches or libraries have you considered?

**Comment by eliboni** [comment]
Well, none actually: I don't think something else exists that allows to do what I'm trying to do without the need of a server-side infrastructure. Being able to easily generate pmtiles for large areas (or all world) and  serve the content so efficiently is something only the brialliant protomps project offers! (am I right?)

Surely, I'm considering maplibre for the future but that would be a major rewrite, with some cons to consider.

**Reply by bdon** [reply]
@eliboni thanks, I agree that there aren't sufficient alternatives for Leaflet for this use case. Is the API as it existed before ideal or can we make it resemble https://maplibre.org/maplibre-gl-js/docs/examples/queryrenderedfeatures/ more? 

We would need to add the caveat that these are "tile features" that do not correspond to what's actually rendered on screen. If a symbolizer filters the tile, the queryFeatures will not be aware of that. So a name that better sets expectations may be `queryTileFeatures` or even `queryFeaturesDebug` to imply this does not do what you expect based on the visual rendering.

**Reply by eliboni** [reply]
@bdon thanks! For my purpose the API as it is if perfect, however I see the value of making it resemble the maplibre one, just to make it easier to switch to maplibre if one wants to do so (or support both). I like the name `queryFeaturesDebug`, since it implies that as a "debug" function it has some  constraint.

**Reply by bdon** [reply]
See proposed `queryTileFeaturesDebug` here: https://github.com/protomaps/protomaps-leaflet/pull/144/files#diff-8aba025ee66c4016dc2af57ce55e4f4ecf88297cad41483db122c3fb485e6725R26

Differences from before: 
* don't add any leaflet UI, you can construct this yourself
* add a extra third parameter `brushSize: number` to determine the effective search area around the point. this is needed for points and lines (polygon use point-in-polygon and are not affected by brushSize)

**Reply by eliboni** [reply]
@dbon downloaded, compiled, tested: it works perfectly. Also, I see how brushSize can be useful. Thanks!

**Reply by bdon** [reply]
@eliboni published as 3.1.0
